File: 2015/day11/day11.java
import java.util.*;

public class day11 {
    public static void main(String[] args) {
        System.out.println("Enter the initial password: ");
        Scanner scanner = new Scanner(System.in);
        String input = scanner.nextLine();

        List<Character> psswd = convertToChar(input);

        do {
            psswd = increment(psswd);
        } while (!validity(psswd));

        System.out.println(stringConverter(psswd));
    }

    static List<Character> increment(List<Character> psswd) {
        int n = psswd.size();
        if (n == 0) {
            System.out.println("No value in password");
            return psswd;
        }

        // 'z' char ASCII value is 122
        // 'a; char ASCII value is 97
        for (int i = n-1; i >= 0; i--) {
            if (psswd.get(i) == 'z') {
                psswd.set(i, 'a');
            } else {
                char newChar = (char) (((int) psswd.get(i)) + 1);
                psswd.set(i, newChar);
                break;
            }
        }
        return psswd;
    }

    // Wrapper method to valid the list of characters
    static boolean validity(List<Character> psswd) {
        return size_validity(psswd)
                && exclude_i_o_l(psswd) // CHANGE ORDER INCREASE PROCESSING TIME
                && three_element_increase(psswd)
                && pair_exist(psswd);
    }

    // If there's 3 elements increase in the list
    static boolean three_element_increase(List<Character> psswd) {
        int n = psswd.size();
        int count = 1;
        int i = 0;
        while (i < n - 1) {
            int dif = (int) psswd.get(i+1) - (int) psswd.get(i);
            if (dif == 1) {
                count++;
                if (count >= 3) {
                    return true;
                }
            } else {
                count = 1;
            }
            i++;
        }
        return false;
    }

    static boolean exclude_i_o_l(List<Character> psswd) {
        final List<Character> exclude = Arrays.asList('i','o','l');
        for (Character character : psswd) {
            if (exclude.contains(character)) {
                return false;
            }
        }
        return true;
    }

    // If there's a non-overlapping pairs
    static boolean pair_exist(List<Character> psswd) {
        int n = psswd.size();
        int pairs = 0;
        int i = 0;

        while (i < n-1) {
            if (psswd.get(i) == psswd.get(i+1)) {
                pairs++;
                i+=2;
            } else {
                i++;
            }
        }

        return pairs >= 2;
    }

    // Check for size
    static boolean size_validity(List<Character> psswd) {
        return psswd.size() == 8;
    }

    // Convert List of String to List to Char
    static List<Character> convertToChar(String str) {
        List<Character> charList = new ArrayList<>();

        for (char ch: str.toCharArray()) {
            charList.add(ch);
        }

        return charList;
    }

    // Convert List of Character to a single String
    static String stringConverter(List<Character> psswd) {
        StringBuilder sb = new StringBuilder();
        for (char c : psswd) {
            sb.append(c);
        }
        return sb.toString();
    }
}

File: 2015/day11/day11.py
def increment(psswd: list[str]):
    n = len(psswd)
    if n == 0:
        return psswd

    for i in range(n-1,-1,-1):
        if psswd[i] == 'z':
            psswd[i] = 'a'
        else:
            psswd[i] = chr(ord(psswd[i]) + 1)
            return psswd

    return psswd

def validity(psswd: list[str]) -> bool:
    return sizeValidity(psswd) and threeElementIncrease(psswd) and excludeiol(psswd) and pairExist(psswd)

def sizeValidity(psswd: list[str]) -> bool:
    return len(psswd) == 8

def threeElementIncrease(psswd: list[str]) -> bool:
    n = len(psswd)
    count = 1
    i = 0
    while i < n - 1:
        dif = ord(psswd[i+1]) - ord(psswd[i])
        if dif == 1:
            count+=1
            if count >= 3:
                return True
        else:
            count = 1

        i += 1

    return False

def excludeiol(psswd: list[str]) -> bool:
    for char in psswd:
        if char in ['i', 'o', 'l']:
            return False
    return True

def pairExist(psswd: list[str]) -> bool:
    n = len(psswd)
    pairs = 0
    i = 0

    while i < n -1:
        if psswd[i] == psswd[i+1]:
            pairs += 1
            i += 2
        else:
            i += 1

    return pairs == 2

def main():
    def getPassword():
        pswd = input("Enter initial password: ").strip()
        return [i for i in pswd]

    try:
        psswd = getPassword()
    except Exception as e:
        print('Input handling error:', e)
        psswd = []

    while True:
        psswd = increment(psswd)

        if not validity(psswd):
            psswd = increment(psswd)
        else:
            break

    print(''.join(psswd))

if __name__ == "__main__":
    main()







